lib/uploader.ts

```diff
@@ -163,34 +171,194 @@ export class Uploader {
 		this._notifiers.push(notifier)
 	}
 
+	/**
+	 * Notify listeners of the upload completion
+	 * @param upload The upload that finished
+	 */
+	private _notifyAll(upload: Upload): void {
+		for (const notifier of this._notifiers) {
+			try {
+				notifier(upload)
+			} catch (error) {
+				// ignore
+			}
+		}
+	}
+
+	/**
+	 * Uploads multiple files or folders while preserving the relative path (if available)
+	 * @param {string} destination The destination path relative to the root folder. e.g. /foo/bar (a file "a.txt" will be uploaded then to "/foo/bar/a.txt")
+	 * @param {Array<File|FileSystemEntry>} files The files and/or folders to upload
+	 * @param {Function} callback Callback that receives the nodes in the current folder and the current path to allow resolving conflicts, all nodes that are returned will be uploaded (if a folder does not exist it will be created)
+	 * @return Cancelable promise that resolves to an array of uploads
+	 *
+	 * @example
+	 * ```ts
+	 * // For example this is from handling the onchange event of an input[type=file]
+	 * async handleFiles(files: File[]) {
+	 *   this.uploads = await this.uploader.bulkUpload('uploads', files, this.handleConflicts)
+	 * }
+	 *
+	 * async handleConflicts(nodes: File[], currentPath: string) {
+	 *   const conflicts = getConflicts(nodes, this.fetchContent(currentPath))
+	 *   if (conficts.length === 0) {
+	 *     // No conflicts so upload all
+	 *     return nodes
+	 *   } else {
+	 *     // Open the conflict picker to resolve conflicts
+	 *     try {
+	 *       const { selected, renamed } = await openConflictPicker(currentPath, conflicts, this.fetchContent(currentPath), { recursive: true })
+	 *       return [...selected, ...renamed]
+	 *     } catch (e) {
+	 *       return false
+	 *     }
+	 *   }
+	 * }
+	 * ```
+	 */
+	bulkUpload(destination: string, files: (File|FileSystemEntry)[], callback?: (nodes: Array<File|Directory>, currentPath: string) => Promise<Array<File|Directory>|false>): PCancelable<Upload[]> {
```

Comment:
Split in multiple lines?
```suggestion
	bulkUpload(
		destination: string,
		files: (File|FileSystemEntry)[],
		callback?: (nodes: Array<File|Directory>, currentPath: string) => Promise<Array<File|Directory>|false>
	): PCancelable<Upload[]> {
```